modules/criteoBidAdapter.js

```diff
@@ -0,0 +1,279 @@
+}
+
+/**
+ * @param {CriteoContext} context
+ * @return {string}
+ */
+function buildCdbUrl(context) {
+  let url = CDB_ENDPOINT;
+  url += '?profileId=' + PROFILE_ID;
+  url += '&av=' + String(ADAPTER_VERSION);
+  url += '&cb=' + String(Math.floor(Math.random() * 99999999999));
+
+  if (context.integrationMode in INTEGRATION_MODES) {
+    url += '&im=' + INTEGRATION_MODES[context.integrationMode];
+  }
+  if (context.debug) {
+    url += '&debug=1';
+  }
+  if (context.noLog) {
+    url += '&nolog=1';
+  }
+
+  return url;
+}
+
+/**
+ * @param {CriteoContext} context
+ * @param {BidRequest[]} bidRequests
+ * @return {*}
+ */
+function buildCdbRequest(context, bidRequests) {
+  let networkId;
+  const request = {
+    publisher: {
+      url: context.url,
+    },
+    slots: bidRequests.map(bidRequest => {
+      networkId = bidRequest.params.networkId || networkId;
+      const slot = {
+        impid: bidRequest.adUnitCode,
+        transactionid: bidRequest.transactionId,
+        auctionId: bidRequest.auctionId,
+        sizes: bidRequest.sizes.map(size => size[0] + 'x' + size[1]),
+      };
+      if (bidRequest.params.zoneId) {
+        slot.zoneid = bidRequest.params.zoneId;
+      }
+      if (bidRequest.params.publisherSubId) {
+        slot.publishersubid = bidRequest.params.publisherSubId;
+      }
+      if (bidRequest.params.nativeCallback) {
+        slot.native = true;
+      }
+      return slot;
+    }),
+  };
+  if (networkId) {
+    request.publisher.networkid = networkId;
+  }
+  return request;
+}
+
+/**
+ * @param {string} id
+ * @param {*} payload
+ * @param {*} callback
+ * @return {string}
+ */
+function createNativeAd(id, payload, callback) {
+  // Store the callback and payload in a global object to be later accessed from the creative
+  window.criteo_prebid_native_slots = window.criteo_prebid_native_slots || {};
+  window.criteo_prebid_native_slots[id] = { callback, payload };
+
+  // The creative is in an iframe so we have to get the callback and payload
+  // from the parent window (doesn't work with safeframes)
+  return `<script type="text/javascript">
+    var win = window;
+    for (var i = 0; i < 10; ++i) {
+      win = win.parent;
+      if (win.criteo_prebid_native_slots) {
+        var responseSlot = win.criteo_prebid_native_slots["${id}"];
+        responseSlot.callback(responseSlot.payload);
+        break;
+      }
+    }
+  </script>`;
+}
+
+/**
+ * @return {boolean}
+ */
+function tryGetCriteoFastBid() {
+  try {
+    const fastBid = localStorage.getItem('criteo_fast_bid');
+    if (fastBid !== null) {
+      eval(fastBid); // eslint-disable-line no-eval
+      return true;
+    }
+  } catch (e) {
+    // Unable to get fast bid
+  }
+  return false;
+}
+
+// Register events to know when Criteo won the bid or timeouted
+let registeredEvents = false;
+function registerEventHandlers() {
+  if (registeredEvents) {
+    return;
+  }
+
+  registeredEvents = true;
+
+  events.on(EVENTS.BID_WON, (bid) => {
```

Comment:
we currently have a policy that bidder adapters cannot consume internal prebid.js events. You'll need to remove these event listeners.

Reply:
- Can I ask what is the rational of this policy? 
- On our side we wanted to track those events to help publishers with their integrations when they encounter a lot of timeouts 
(some of them have inconsistencies in their timeouts configuration between the one passed in requestBids call and the one in the global setTimeout)
- As documented here (http://prebid.org/dev-docs/publisher-api-reference.html#module_pbjs.onEvent), can we replace the call to "events.on" by onEvent ?
- If not what would be the right alternative way to collect those informations ?

Reply:
Reason is that we don't want vendors inspecting other vendor bids. I realize that you are not doing that but we want to make this preventable in the code. 

The correct way is not fully implemented yet, but you could start with `spec.onTimeout` function.
http://prebid.org/dev-docs/bidder-adaptor.html#registering-on-timeout

Other events can be added to the spec in a separate PR if desired. Thanks